proposals/context-parameters.md

```diff
@@ -0,0 +1,571 @@
+# Context parameters
+
+* **Type**: Design proposal
+* **Author**: Alejandro Serrano
+* **Contributors**: Marat Akhin, Nikita Bobko, Ilya Gorbunov, Mikhail Zarechenskii, Denis Zharkov
+* **Discussion**: [KEEP-367](https://github.com/Kotlin/KEEP/issues/367)
+
+## Abstract
+
+This is an updated proposal for [KEEP-259](https://github.com/Kotlin/KEEP/issues/259), formerly known as _context receivers_. The new design addresses the issues raised by the users of the prototype implementing that previous iteration of the design and across the community at large. 
+
+This document is not (yet) formally a KEEP, since it lacks some of the technical elements. Those are going to be provided at a later time, but we thought it would be interesting to open the discussion even if the design is not fully formalized.
+
+### Summary of changes from the [previous proposal](https://github.com/Kotlin/KEEP/issues/259)
+
+1. Introduction of named context parameters,
+2. Context receivers are dropped,
+3. Removal of `this@Type` syntax, introduction of `implicit<A>()`,
+4. Contexts are not allowed in constructors,
+5. Callable references resolve their context arguments eagerly,
+6. Context-in-classes are dropped.
+
+## Table of contents
+
+* [Abstract](#abstract)
+  * [Summary of changes from the previous proposal](#summary-of-changes-from-the-previous-proposal)
+* [Table of contents](#table-of-contents)
+* [Members with context parameters](#members-with-context-parameters)
+* [Standard library support](#standard-library-support)
+  * [Reflection](#reflection)
+* [Simulating receivers](#simulating-receivers)
+* [Use cases](#use-cases)
+  * [As implicits](#as-implicits)
+  * [As scopes](#as-scopes)
+  * [For extending DSLs](#for-extending-dsls)
+  * [Context-oriented dispatch / externally-implemented interface / type classes](#context-oriented-dispatch--externally-implemented-interface--type-classes)
+  * [Dependency injection](#dependency-injection)
+* [Callable references](#callable-references)
+* [Context and classes](#context-and-classes)
+* [Technical design](#technical-design)
+  * [Syntax](#syntax)
+  * [Extended resolution algorithm](#extended-resolution-algorithm)
+  * [Extended type inference algorithm](#extended-type-inference-algorithm)
+  * [JVM ABI and Java compatibility](#jvm-abi-and-java-compatibility)
+* [Q\&A about design decisions](#qa-about-design-decisions)
+* [Acknowledgments](#acknowledgments)
+
+
+## Members with context parameters
```

Comment:
"Members" makes an impression that context parameters are allowed only in member functions and properties, i.e. ones inside a class or an object.